packages/@aws-cdk/aws-iotevents/lib/state.ts

```diff
@@ -28,21 +28,46 @@ export class State {
    */
   public readonly stateName: string;
 
+  private transitionEvents: ITransitionEvent[] = []
+
   constructor(private readonly props: StateProps) {
     this.stateName = props.stateName;
   }
 
   /**
-   * Return the state property JSON
+   * Add a transition event to the state.
+   *
+   * @param transitionEvent the transition event that triggered if condition is evaluated to TRUE
+   */
+  public transitionTo(transitionEvent:ITransitionEvent) {
+    this.transitionEvents.push(transitionEvent);
+  }
+
+  /**
+   * Return the JSON of the states that is transited to.
+   * This function is called recursively and collect the states.
    *
    * @internal
    */
-  public _toStateJson(): CfnDetectorModel.StateProperty {
+  public _getStatesJson(states: CfnDetectorModel.StateProperty[] = []): CfnDetectorModel.StateProperty[] {
     const { stateName, onEnter } = this.props;
-    return {
+
+    if (states.some(s => s.stateName === stateName)) {
+      return states;
+    }
+
+    const newStates: CfnDetectorModel.StateProperty[] = [...states, {
       stateName,
+      onInput: {
+        transitionEvents: this.transitionEvents.length > 0 ?
+          getTransitionEventJson(this.transitionEvents) : undefined,
+      },
       onEnter: onEnter && { events: getEventJson(onEnter) },
-    };
+    }];
+
+    return this.transitionEvents.reduce((acc, transitionEvent) => {
+      return transitionEvent.nextState._getStatesJson(acc);
+    }, newStates);
```

Comment:
I have to say, I don't understand why do we have to call this method recursively? Why does translating all `this.transitionEvents` Events isn't good enough?

Reply:
OK.
IoT Events can construct state dependency graph, for example the following image:
![image](https://user-images.githubusercontent.com/11013683/152146897-13d788fb-a499-4022-a67f-5c9abcba669b.png)

When now, `firstState` has only one transition `firstState_to_secondState` and can collect only one state. Therefore, it is necessary to recursively collect states by tracing transitions from the entry state.